pkg/policy/rule.go

```diff
@@ -35,54 +35,93 @@ func (r *rule) String() string {
 	return fmt.Sprintf("%v", r.EndpointSelector)
 }
 
-func (policy *L4Filter) addEndpoints(endpoints []api.EndpointSelector) bool {
-
-	if len(policy.Endpoints) == 0 && len(endpoints) > 0 {
-		log.WithFields(logrus.Fields{
-			logfields.EndpointSelector: endpoints,
-			"policy":                   policy,
-		}).Debug("skipping L4 filter as the endpoints are already covered.")
-		return true
-	}
-
-	if len(policy.Endpoints) > 0 && len(endpoints) == 0 {
-		log.WithFields(logrus.Fields{
-			logfields.EndpointSelector: endpoints,
-			"policy":                   policy,
-		}).Debug("new L4 filter applies to all endpoints, making the policy more permissive.")
-		policy.Endpoints = nil
-	}
-
-	policy.Endpoints = append(policy.Endpoints, endpoints...)
-	return false
-}
-
+// mergeL4IngressPort merges all rules which share the same port & protocol that
+// select a given set of endpoints. It updates the L4Filter mapped to by the specified
+// port and protocol with the contents of the provided PortRule. If the rule
+// being merged has conflicting L7 rules with those already in the provided
+// L4PolicyMap for the specified port-protocol tuple, it returns an error.
 func mergeL4IngressPort(ctx *SearchContext, endpoints []api.EndpointSelector, r api.PortRule, p api.PortProtocol,
 	proto api.L4Proto, ruleLabels labels.LabelArray, resMap L4PolicyMap) (int, error) {
 
 	key := p.Port + "/" + string(proto)
-	filter, ok := resMap[key]
+	existingFilter, ok := resMap[key]
 	if !ok {
 		resMap[key] = CreateL4IngressFilter(endpoints, r, p, proto, ruleLabels)
 		return 1, nil
 	}
-	l4Filter := CreateL4IngressFilter(endpoints, r, p, proto, ruleLabels)
-	if l4Filter.L7Parser != "" {
-		if filter.L7Parser == "" {
-			filter.L7Parser = l4Filter.L7Parser
-		} else if l4Filter.L7Parser != filter.L7Parser {
-			ctx.PolicyTrace("   Merge conflict: mismatching parsers %s/%s\n", l4Filter.L7Parser, filter.L7Parser)
-			return 0, fmt.Errorf("Cannot merge conflicting L7 parsers (%s/%s)", l4Filter.L7Parser, filter.L7Parser)
+
+	// Create a new L4Filter based off of the arguments provided to this function
+	// for merging with the filter which is already in the policy map.
+	filterToMerge := CreateL4IngressFilter(endpoints, r, p, proto, ruleLabels)
+
+	// Handle cases where filter we are merging new rule with, or new rule itself
+	// allows all traffic on L3.
+	//
+	// Existing filter selects all endpoints, so don't add new endpoints.
+	if len(existingFilter.Endpoints) == 0 && len(filterToMerge.Endpoints) > 0 {
```

Comment:
Is `len(Endpoints) == 0` reliable to determine the wildcard? What if Endpoints has an entry with the `WildcardEndpointSelector`?

Reply:
`len(Endpoints) == 0` represents all endpoints being selected:  https://github.com/cilium/cilium/blob/9571490cbb45bfb93fa230a2c7a6101b03101a56/pkg/policy/l4.go#L88-L90 

This is confusing since we have the wildcard endpoint selector to represent all endpoints being selected. This should probably be changed to just having the slice contain the WildcardEndpointSelector and then having some helper function for L4Filter called `SelectsAllEndpoints`, which checks if `len(endpoints) == 1` and `endpoints[0] == api.WildcardEndpointSelector`.

Reply:
Will also add a unit test with the `WildcardEndpointSelector`.

Reply:
If you checked that we are never passing `[]api.EndpointSelector{WildcardEndpointSelector}` in then the code is fine as-is. I agree that we should clean it up in the next feature branch.

Reply:
Adding explicit allow-all at L3 (i.e., with `WildCardEndpointSelector`) caused breakage with the logic I had in place, so I added a function to check if `EndpointSelectorSlice`selects all endpoints, then changed the list of `Endpoints` in `L4Filter` to be an `EndpointSelectorSlice`. I added new unit tests to validate this behavior, and it now behaves just like the case where if no EndpointSelector was provided at all. Let me know what you think.